pyterrier/apply_base.py

```diff
@@ -68,6 +69,39 @@ def transform(self, res):
                 raise ValueError("Cannot apply add_ranks in pt.apply.by_query - " + suffix) from ke
         rtr = pd.concat(query_dfs)
         return rtr
+    
+class ApplyIterForEachQuery(ApplyTransformerBase):
+    def __init__(self, fn,  *args, add_ranks=False, batch_size=None, **kwargs):
+        """
+            Arguments:
+             - fn (Callable): Takes as input a panda Series for a row representing that document, and returns the new float doument score 
+        """
+        super().__init__(fn, *args, **kwargs)
+        self.add_ranks = add_ranks
+        self.batch_size = batch_size
+        assert not kwargs.get("verbose", False), "verbose not supported"
+    
+    def __repr__(self):
+        return "pt.apply.by_query()"
+    
+    def transform_iter(self, input):
+        import more_itertools
+        assert not self.add_ranks
+        input = more_itertools.peekable(input)
+        try:
+            thisqid = input.peek()["qid"]
+        except StopIteration:
+            # input iterable was empty
+            yield from [] 
+        batch = []
+        for row in input:
+            if row["qid"] != thisqid or self.batch_size is not None and len(batch) == self.batch_size:
```

Comment:
I think we can use `itertools.groupby` here to simplify a bit

Reply:
groups returned by groupby are iterators, so we couldnt have batch_size support... I've used more_itertools.ichunked to resolve this.